### PR TITLE
Added support to undo & redo file operations

### LINKED ISSUE #1975: Add Undo, Redo functionallity
**Is your feature request related to a problem? Please describe.**
It's sometimes annoying when you copy file(s) to a folder that you didn't intend to. Same with moving/deleting items.

**Describe the solution you'd like**
ctrl+z, ctrl+y to undo/redo filesystem operations

### PR DESCRIPTION
**Pull request by d2dyno1**
This PR features complete Undo, Redo functionality. This PR is compatible with newest commit without merge issues. Please be aware that there still might be some bugs out there that sild in unnoticed.

Resolves #1975
Closes #2250

### CODE REVIEW

**Comment by d2dyno1** [comment]
The PR is Ready to be reviewed

File: Files/Filesystem/StorageHistory/Helpers/StorageHistoryHelpers.cs [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+            if (CanUndo())
+            {
+                App.StorageHistoryIndex--;
+                int index = EnumerableHelpers.FitBounds(App.StorageHistoryIndex, App.StorageHistory.Count);
+                return await this._storageHistoryOperations.Undo(App.StorageHistory[index]);
+            }
+            return Status.InProgress;
+        }
+
+        public async Task<Status> Redo()
```
Comment (yaichenbaum) [inline Files/Filesystem/StorageHistory/Helpers/StorageHistoryHelpers.cs:37]: I am not sure we want to implement a redo feature.

Comment (d2dyno1) [reply Files/Filesystem/StorageHistory/Helpers/StorageHistoryHelpers.cs:37]: Well the default Windows 10 File Explorer has the undo feature.
Why discard it if it's already implemented and working. I'm certain a lot of users (including me) will find the Redo useful

File: Files/Helpers/Convert/StorageItemHelpers.cs [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+using System.Threading.Tasks;
+using Windows.Storage;
+
+namespace Files.Helpers
+{
+    public static class StorageItemHelpers
+    {
+        public static async Task<IStorageItem> ToStorageItem(this string path)
+        {
+            StorageFolder folder = await StorageFolder.GetFolderFromPathAsync(Path.GetDirectoryName(path));
```
Comment (gave92) [inline Files/Helpers/Convert/StorageItemHelpers.cs:14]: This is likely to throw exceptions, should probably be wrapped with `FilesystemTasks.Wrap()` and explicit error handling should be added whenever it's used

**Review by gave92 — DISMISSED** [review]
Hi @d2dyno1 this is coming along nicely!

Some testing resulted in the following issues:

- crash deleting from recycle bin
2020-11-16 23:37:51.9637|ERROR|Files.App|Accesso negato. (Exception from HRESULT: 0x80070005 (E_ACCESSDENIED))|System.UnauthorizedAccessException: Accesso negato. (Exception from HRESULT: 0x80070005 (E_ACCESSDENIED))
   at Files.Helpers.StorageItemHelpers.ToStorageItemCollection(IEnumerable`1 listedItems)
   at Files.Views.Pages.ModernShellPage.KeyboardAccelerator_Invoked(KeyboardAccelerator sender, KeyboardAcceleratorInvokedEventArgs args)
   at System.Threading.WinRTSynchronizationContextBase.Invoker.InvokeCore()
- crash on drag and drop after hovering on a folder (and so navigating to it)
2020-11-16 23:39:21.3476|ERROR|Files.App|Object reference not set to an instance of an object.|System.NullReferenceException: Object reference not set to an instance of an object.
   at Files.Filesystem.FilesystemHelpers.MoveItemAsync(IStorageItem source, String destination, Boolean registerHistory)
   at Files.Filesystem.FilesystemHelpers.MoveItemsFromClipboard(DataPackageView packageView, String destination, Boolean registerHistory)
   at Files.Filesystem.FilesystemHelpers.PerformOperationTypeAsync(DataPackageOperation operation, DataPackageView packageView, String destination, Boolean registerHistory)
   at Files.BaseLayout.List_Drop(Object sender, DragEventArgs e)
   at System.Threading.WinRTSynchronizationContextBase.Invoker.InvokeCore()
- cltr-x copies instead of cutting
- cannot copy paste shortcut (.lnk) files
- crash pasting file to windows folder
2020-11-16 23:46:15.5820|ERROR|Files.App|Object reference not set to an instance of an object.|System.NullReferenceException: Object reference not set to an instance of an object.
   at Files.Filesystem.FilesystemOperations.CopyAsync(IStorageItem source, String destination, IProgress`1 progress, IProgress`1 errorCode, CancellationToken cancellationToken)
   at Files.Filesystem.FilesystemHelpers.CopyItemAsync(IStorageItem source, String destination, Boolean registerHistory)
   at Files.Filesystem.FilesystemHelpers.CopyItemsFromClipboard(DataPackageView packageView, String destination, Boolean registerHistory)
   at Files.Filesystem.FilesystemHelpers.PerformOperationTypeAsync(DataPackageOperation operation, DataPackageView packageView, String destination, Boolean registerHistory)
   at Files.Interacts.Interaction.PasteItem()
   at System.Threading.WinRTSynchronizationContextBase.Invoker.InvokeCore()
- does not copy files to usb connected android phone anymore
- crash when (this is difficult XD):
1) copypaste a file to its same folder
2) rename new file to something else
3) ctrlz
4) ctrlz
5) ctrly
2020-11-16 23:35:48.9522|ERROR|Files.App|Impossibile trovare il file specificato. (Exception from HRESULT: 0x80070002)|System.IO.FileNotFoundException: Impossibile trovare il file specificato. (Exception from HRESULT: 0x80070002)
   at Files.Helpers.StorageItemHelpers.ToStorageItem(String path)
   at Files.Filesystem.FilesystemHistory.StorageHistoryOperations.Redo(IStorageHistory history)
   at Files.Filesystem.FilesystemHistory.StorageHistoryHelpers.Redo()
   at Files.Views.Pages.ModernShellPage.KeyboardAccelerator_Invoked(KeyboardAccelerator sender, KeyboardAcceleratorInvokedEventArgs args)
   at System.Threading.WinRTSynchronizationContextBase.Invoker.InvokeCore()

File: Files/BaseLayout.cs [hunk 3]
```diff
@@ -33,6 +33,8 @@ namespace Files
     /// </summary>
     public abstract class BaseLayout : Page, INotifyPropertyChanged
     {
+        private IFilesystemHelpers _filesystemHelpers;
```
[the remark this answers is not in the document]
Comment (tsvietOK) [reply Files/BaseLayout.cs:36]: camelCase for private members, PascalCase for public.

**Comment by d2dyno1** [comment]
Looks like I'll be busy...

**Comment by d2dyno1** [comment]
> * crash when (this is difficult XD):
> 
> 1. copypaste a file to its same folder
> 2. rename new file to something else
> 3. ctrlz
> 4. ctrlz
> 5. ctrly

Are you trying to cast spells with this combination 😆

**Comment by d2dyno1** [comment]
Here are all the changes requested:
- Crash deleting from recycle bin
- Crash on drag and drop after hovering on a folder (and so navigating to it)
- Cltr-x copies instead of cutting
- Cannot copy paste shortcut (.lnk) files
- Crash pasting file to windows folder
- Does not copy files to usb connected android phone anymore
- Witchcraft combination
- Remame Status to something more specific
- Use FilesystemTasks.Wrap() on ToStorageItem()

I'll report progress on them once they're completed

**Comment by d2dyno1** [comment]
> crash pasting file to windows folder
@gave92 Hmmm... It doesn't seem to crash for me

**Comment by gave92** [comment]
@d2dyno1 I'll check again 👍

**Comment by gave92** [comment]
Thanks for your work on this, here some updates on the issues:

- crash pasting file to windows folder
-> I can confirm this, pasting to "C:\Program Files (x86)" does the same
- "magic combo"
copypaste a file to its same folder
rename new file to something else
ctrlz
ctrlz
ctrly
-> no crash anymore but the "ctrly" does nothing (should redo copy-paste)
- quickly pressing ctrlz + ctrly often results in a crash
copypaste file
*quickly* press ctrlz and ctrly
2020-11-17 19:29:54.0065|ERROR|Files.App|Object reference not set to an instance of an object.|System.NullReferenceException: Object reference not set to an instance of an object.
   at Files.Filesystem.FilesystemOperations.DeleteAsync(IStorageItem source, IProgress`1 progress, IProgress`1 errorCode, Boolean permanently, CancellationToken cancellationToken)
   at Files.Filesystem.FilesystemHistory.StorageHistoryOperations.Redo(IStorageHistory history)
   at Files.Filesystem.FilesystemHistory.StorageHistoryHelpers.Redo()
   at Files.Views.Pages.ModernShellPage.KeyboardAccelerator_Invoked(KeyboardAccelerator sender, KeyboardAcceleratorInvokedEventArgs args)
   at System.Threading.WinRTSynchronizationContextBase.Invoker.InvokeCore()

File: Files/BaseLayout.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (tsvietOK) [reply Files/BaseLayout.cs:36]: @d2dyno1 Code style for this project you can find [here](https://github.com/files-community/Files/wiki/Code-style).

File: Files/Controllers/TerminalController.cs [hunk 4]
```diff
@@ -70,8 +70,8 @@ private async Task LoadAsync()
 
         public async void Init()
         {
-            await LoadAsync();
-            await GetInstalledTerminalsAsync();
+            //await LoadAsync();
+            //await GetInstalledTerminalsAsync();
```
Comment (yaichenbaum) [inline Files/Controllers/TerminalController.cs:74]: Is this change required for this pull request?

Comment (d2dyno1) [reply Files/Controllers/TerminalController.cs:74]: I commented it out because it wouldn't let me compile. I'll revert the comments once the PR is done

File: Files/BaseLayout.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (d2dyno1) [reply Files/BaseLayout.cs:36]: Hmm ok, I'll change the variable names. I like using prefixes to know which variable I'm accessing. Without the prefixes it'd get confusing pretty fast

File: Files/Filesystem/StorageFileHelpers/IStorageItemWithPath.cs [hunk 5]
```diff
@@ -2,7 +2,8 @@
 
 namespace Files.Filesystem
 {
-    public interface IStorageItemWithPath
+    public interface IStorageItemWithPath // TODO: Maybe use here : IStorageItem instead of declaring a variable,
```
Comment (d2dyno1) [inline Files/Filesystem/StorageFileHelpers/IStorageItemWithPath.cs:5]: Hmm...

File: Files/Helpers/Convert/StorageItemHelpers.cs [hunk 6]
```diff
@@ -12,28 +12,46 @@ namespace Files.Helpers
...
             {
-                return fsFileResult.Result;
+                if (!string.IsNullOrWhiteSpace(fsFileResult.Result.Path))
+                    return fsFileResult.Result;
+                else
+                {
+                    FilesystemResult<StorageFileWithPath> fsFileWithPathResult = await FilesystemTasks.Wrap(() => StorageFileExtensions.DangerousGetFileWithPathFromPathAsync(path, rootFolder));
+
+                    if (fsFileWithPathResult)
+                        return /* fsFileWithPathResult.Result */ null; // Could be done if IStorageItemWithPath implemented IStorageItem
```
Comment (d2dyno1) [inline Files/Helpers/Convert/StorageItemHelpers.cs:30]: :confused:

File: Files/Helpers/Convert/StorageItemHelpers.cs [hunk 7]
```diff
@@ -12,28 +12,46 @@ namespace Files.Helpers
...
             {
-                return fsFolderResult.Result;
+                 if (!string.IsNullOrWhiteSpace(fsFolderResult.Result.Path))
+                    return fsFolderResult.Result;
+                else
+                {
+                    FilesystemResult<StorageFolderWithPath> fsFolderWithPathResult = await FilesystemTasks.Wrap(() => StorageFileExtensions.DangerousGetFolderWithPathFromPathAsync(path, rootFolder));
+
+                    if (fsFolderWithPathResult)
+                        return /* fsFolderWithPathResult.Result; */ null; // Could be done if IStorageItemWithPath implemented IStorageItem
```
Comment (d2dyno1) [inline Files/Helpers/Convert/StorageItemHelpers.cs:45]: :confused:

File: Files/Helpers/Convert/StorageItemHelpers.cs (hunk 6 above)
Comment (d2dyno1) [reply Files/Helpers/Convert/StorageItemHelpers.cs:30]: Is implementing IStorageItem a possibility. Or would it have to work different way?

File: Files/Filesystem/FilesystemOperations/FilesystemOperations.cs [hunk 8]
```diff
@@ -250,25 +250,39 @@ public async Task<IStorageHistory> MoveAsync(IStorageItem source, string destina
...
+            bool deleteFromRecycleBin = await _recycleBinHelpers.IsRecycleBinItem(source);
 
             FilesystemResult fsResult = FilesystemErrorCode.ERROR_INPROGRESS;
             FilesystemItemType itemType = source.IsOfType(StorageItemTypes.File) ? FilesystemItemType.File : FilesystemItemType.Directory;
 
             errorCode?.Report(fsResult);
             progress?.Report(0.0f);
 
-            if (source.IsOfType(StorageItemTypes.File))
+            if (source is IStorageItemWithPath sourceWithPath)
```
Comment (d2dyno1) [inline Files/Filesystem/FilesystemOperations/FilesystemOperations.cs:261]: I know this will never be executed

**Comment by d2dyno1** [comment]
Ok, I got around my issue, but it'd be way cleaner if `IStorageItemWithPath` implemented `IStorageItem`

**Comment by gave92** [comment]
@d2dyno1 glad you solved it. But making IStorageItemWithPath implement IStorageItem is something that we can totally consider.

File: Files/Filesystem/StorageHistory/Helpers/StorageHistoryHelpers.cs [hunk 9]
```diff
@@ -8,62 +12,70 @@ public class StorageHistoryHelpers : IDisposable
     {
         #region Private Members
 
-        private IStorageHistoryOperations _storageHistoryOperations;
+        private IStorageHistoryOperations storageHistoryOperations;
+
+        private Queue<Func<Task<ReturnResult>>> operationsQueue = new Queue<Func<Task<ReturnResult>>>();
```
Comment (d2dyno1) [inline Files/Filesystem/StorageHistory/Helpers/StorageHistoryHelpers.cs:17]: @gave92 I tried to throttle the tasks but I'm not entirely sure how to do it

File: Files/Filesystem/StorageHistory/Helpers/StorageHistoryHelpers.cs [hunk 10]
```diff
@@ -33,22 +32,50 @@ public async Task Undo()
...
-                operationsQueue.Enqueue(new Func<Task<ReturnResult>>(() => this.storageHistoryOperations.Undo(App.StorageHistory[index])));
+                try
+                {
+                    await App.SemaphoreSlim.WaitAsync(0);
+
+                    App.StorageHistoryIndex--;
+                    int index = EnumerableHelpers.FitBounds(App.StorageHistoryIndex, App.StorageHistory.Count);
+                    await this.storageHistoryOperations.Undo(App.StorageHistory[index]);
+
+                    await Task.Delay(10000); // Test
```
Comment (d2dyno1) [inline Files/Filesystem/StorageHistory/Helpers/StorageHistoryHelpers.cs:43]: For testing purposes

**Comment by d2dyno1** [comment]
Fixed Spamming Ctrl+Z, Ctrl+Y with a little help from @gave92 😉 🙌

**Comment by gave92** [comment]
Hi @d2dyno1, I can confirm that the issue "quickly pressing ctrlz + ctrly often results in a crash" is fixed.
Also I'm not getting a crash anymore when "pasting file to windows folder"
But:
- copy pasting a folder to the same location causes issues: folder is not copied (should be duplicated with a "(2)" at the end) and ctrl-z deletes the origin folder
- copy pasting a folder "inside itself" and pressing ctrl-z causes a crash

**Comment by d2dyno1** [comment]
> Here are all the changes requested:
> 
> * [x]  Crash deleting from recycle bin
> * [x]  Dropping dragged items to a folder causes a crash (and so navigating to it) ******
> * [x]  CTRL+X copies instead of cutting
> * [x]  Cannot copy paste shortcut (.lnk) files
> * [x]  Crash pasting items to /Windows (and /Program Files (x86)) directories
> * [x]  Cannot delete files from usb connected android phone
> * [x]  Witchcraft combination
> * [x]  Remame Status to something more specific
> * [x]  Use FilesystemTasks.Wrap() on ToStorageItem()
> * [x]  Quickly pressing ctrlz + ctrly often results in a crash
> * [x]  Copy pasting a folder to the same location does not duplicate the folder
> * [x]  Copy pasting a folder "inside itself" and pressing ctrl-z causes a crash
> 
> I'll report progress on them once they're completed
> 
> ***** - TBD

@gave92 looks like everything is done :eyes:

**Comment by gave92** [comment]
This `source.Path == Path.GetDirectoryName(destination)` only works if you copy the folder "immediately" inside itself. If you copy the folder at least "two levels down" you'll still have issues (untested)

**Comment by d2dyno1** [comment]
Good catch!

**Comment by d2dyno1** [comment]
Now it should be ok

**Comment by gave92** [comment]
Nope sorry I still get an exception when I press ctrl-z (after pasting a folder inside itself)

2020-11-20 20:55:15.1122|ERROR|Files.App|Object reference not set to an instance of an object.|System.NullReferenceException: Object reference not set to an instance of an object.
   at Files.Helpers.RecycleBinHelpers.<>c__DisplayClass5_0.<IsRecycleBinItem>b__0(ShellFileItem shellItem)
   at System.Linq.Enumerable.Any[TSource](IEnumerable`1 source, Func`2 predicate)
   at Files.Helpers.RecycleBinHelpers.IsRecycleBinItem(IStorageItem item)
   at Files.Filesystem.FilesystemHelpers.DeleteItemsAsync(IEnumerable`1 source, Boolean showDialog, Boolean permanently, Boolean registerHistory)
   at Files.Filesystem.FilesystemHistory.StorageHistoryOperations.Undo(IStorageHistory history)
   at Files.Filesystem.FilesystemHistory.StorageHistoryHelpers.Undo()
   at Files.Views.Pages.ModernShellPage.KeyboardAccelerator_Invoked(KeyboardAccelerator sender, KeyboardAcceleratorInvokedEventArgs args)
   at System.Threading.WinRTSynchronizationContextBase.Invoker.InvokeCore()

**Comment by d2dyno1** [comment]
Oh, probably because it returned invalid value, gonna fix it, gimme a momento

**Comment by gave92** [comment]
More issues found :)

Pressing ctrl-z after copying a hosrtcut (.lnk) couses this:
2020-11-20 21:33:44.8706|ERROR|Files.App|Object reference not set to an instance of an object.|System.NullReferenceException: Object reference not set to an instance of an object.
   at Files.Filesystem.FilesystemHelpers.GetItemSize(IStorageItem item)
   at Files.Filesystem.FilesystemOperations.CopyAsync(IStorageItem source, String destination, IProgress`1 progress, IProgress`1 errorCode, CancellationToken cancellationToken)
   at Files.Filesystem.FilesystemOperations.MoveAsync(IStorageItem source, String destination, IProgress`1 progress, IProgress`1 errorCode, CancellationToken cancellationToken)
   at Files.Filesystem.FilesystemHelpers.MoveItemAsync(IStorageItem source, String destination, Boolean registerHistory)
   at Files.Filesystem.FilesystemHistory.StorageHistoryOperations.Undo(IStorageHistory history)
   at Files.Filesystem.FilesystemHistory.StorageHistoryHelpers.Undo()
   at Files.Views.Pages.ModernShellPage.KeyboardAccelerator_Invoked(KeyboardAccelerator sender, KeyboardAcceleratorInvokedEventArgs args)
   at System.Threading.WinRTSynchronizationContextBase.Invoker.InvokeCore()

**Comment by d2dyno1** [comment]
> Shortcut items now work with Undo, Redo but... (read description)
Untested, I ran out of time today

**Comment by d2dyno1** [comment]
Got one more commit for today 😌

**Comment by gave92** [comment]
@d2dyno1 undo/redo seems to work ok for shortcuts 👍 except for:
- crash on sending a shortcut item to recycle bin
2020-11-21 17:09:45.7188|ERROR|Files.App|Unsupported data type. (Exception from HRESULT: 0x8007065E)|System.Exception: Unsupported data type. (Exception from HRESULT: 0x8007065E)
   at System.Runtime.InteropServices.WindowsRuntime.IMap`2.Insert(K key, V value)
   at System.Runtime.InteropServices.WindowsRuntime.MapToDictionaryAdapter.Insert[K,V](IMap`2 _this, K key, V value)
   at System.Runtime.InteropServices.WindowsRuntime.MapToDictionaryAdapter.Add[K,V](K key, V value)
   at Files.Filesystem.FilesystemOperations.DeleteAsync(PathWithType source, IProgress`1 progress, IProgress`1 errorCode, Boolean permanently, CancellationToken cancellationToken)
   at Files.Filesystem.FilesystemHelpers.DeleteItemsAsync(IEnumerable`1 source, Boolean showDialog, Boolean permanently, Boolean registerHistory)
   at Files.Views.Pages.ModernShellPage.KeyboardAccelerator_Invoked(KeyboardAccelerator sender, KeyboardAcceleratorInvokedEventArgs args)
   at System.Threading.WinRTSynchronizationContextBase.Invoker.InvokeCore()

**Comment by d2dyno1** [comment]
Hmm... the exception is thrown here:
image
That's not the case with original code, but the logic used is the same 🤔

**Comment by d2dyno1** [comment]
:eyes:

**Comment by d2dyno1** [comment]
I hope that's all, and this commit can be finally fully reviewed @gave92  @yair100  @tsvietOK

File: Files/DataModels/ReturnResult.cs [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+        /// </summary>
+        BadArgumentException = 5,
+
+        /// <summary>
+        /// Informs that operation provided/returned value is null
+        /// </summary>
+        NullException = 6,
+
+        /// <summary>
+        /// Infoms that operation tried to access restricted resources
```
Comment (yaichenbaum) [inline Files/DataModels/ReturnResult.cs:47]:
```suggestion
        /// Informs that operation tried to access restricted resources
```

**Review by yair100 — DISMISSED** [review]

**Comment by tsvietOK** [comment]
Bug steps:
1. Delete file to Recycle bin
2. Clear Recycle bin
3. Press Ctrl+Z.
4. Press Ctrl+Y.
5. Crash

**Comment by d2dyno1** [comment]
I'm aware it might have not fixed all issues

**Comment by d2dyno1** [comment]
Should be alright now, @tsvietOK

File: Files/Filesystem/StorageHistory/Helpers/StorageHistoryHelpers.cs [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+        public StorageHistoryHelpers(IStorageHistoryOperations storageHistoryOperations)
+        {
+            this.storageHistoryOperations = storageHistoryOperations;
+        }
+
+        #endregion
+
+        #region Undo, Redo
+
+        public async Task<ReturnResult> Undo()
```
Comment (tsvietOK) [inline Files/Filesystem/StorageHistory/Helpers/StorageHistoryHelpers.cs:26]: Rename to `TryUndo`.

File: Files/Filesystem/StorageHistory/Helpers/StorageHistoryHelpers.cs [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+                finally
+                {
+                    App.SemaphoreSlim.Release();
+                }
+            }
+
+            return ReturnResult.Cancelled;
+        }
+
+        public async Task<ReturnResult> Redo()
```
Comment (tsvietOK) [inline Files/Filesystem/StorageHistory/Helpers/StorageHistoryHelpers.cs:48]: Rename to `TryRedo`.

File: Files/Filesystem/StorageHistory/Helpers/StorageHistoryHelpers.cs [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+
+        #endregion
+
+        #region Undo, Redo
+
+        public async Task<ReturnResult> Undo()
+        {
+            if (CanUndo())
+            {
+                if (!(await App.SemaphoreSlim.WaitAsync(0)))
```
Comment (tsvietOK) [inline Files/Filesystem/StorageHistory/Helpers/StorageHistoryHelpers.cs:30]: [Code style 11](https://github.com/files-community/Files/wiki/Code-style#11), check other changes too.

File: Files/Interacts/Interaction.cs [hunk 15]
```diff
@@ -915,32 +883,32 @@ public async Task<bool> RenameFileItemAsync(ListedItem item, string oldName, str
...
-                                .OnSuccess(t => t.RenameAsync(newName, NameCollisionOption.ReplaceExisting).AsTask());
+                            renamed = await this.AssociatedInstance.FilesystemViewModel.GetFileFromPathAsync(item.ItemPath)
+                                .OnSuccess(async (t) => await this.FilesystemHelpers.RenameAsync(t, newName, NameCollisionOption.ReplaceExisting, true));
                         }
                     }
                 }
                 if (renamed)
                 {
-                    AssociatedInstance.NavigationToolbar.CanGoForward = false;
+                    this.AssociatedInstance.NavigationToolbar.CanGoForward = false;
```
Comment (tsvietOK) [inline Files/Interacts/Interaction.cs:911]: `this` is optional operator.

**Review by tsvietOK — CHANGES_REQUESTED** [review]

**Comment by d2dyno1** [comment]
50th comment!

File: Files/Extensions/LinqExtensions.cs [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+
+        /// <summary>
+        /// Enumerates through <paramref name="collection"/> of elements and executes <paramref name="func"/> and returns <typeparamref name="T2"/>
+        /// </summary>
+        /// <typeparam name="T1">Element of <paramref name="collection"/></typeparam>
+        /// <typeparam name="T2">The result of <paramref name="func"/></typeparam>
+        /// <param name="collection">The collection to enumerate through</param>
+        /// <param name="func">The func to take every element</param>
+        /// <returns>Result of <paramref name="func"/> of every enumerated item</returns>
+        internal static IEnumerable<T2> ForEach<T1, T2>(this IEnumerable<T1> collection, Func<T1, T2> func)
```
Comment (tsvietOK) [inline Files/Extensions/LinqExtensions.cs:28]: Unused. Is it needed?

**Review by tsvietOK — DISMISSED** [review]

File: Files/Filesystem/StorageFileHelpers/FilesystemResult.cs [hunk 17]
```diff
@@ -8,22 +8,23 @@ namespace Files.Filesystem
...
         ERROR_INUSE = 8,
         ERROR_NAMETOOLONG = 16,
         ERROR_ALREADYEXIST = 32,
         ERROR_NOTAFOLDER = 64,
-        ERROR_NOTAFILE = 128
+        ERROR_NOTAFILE = 128,
+        ERROR_INPROGRESS = 256
     }
 
     public static class FilesystemErrorCodeExtensions
```
[the remark this answers is not in the document]
Comment (d2dyno1) [reply Files/Filesystem/StorageFileHelpers/FilesystemResult.cs:23]: It is actually used in 12 instances

File: Files/Extensions/LinqExtensions.cs (hunk 16 above)
Comment (d2dyno1) [reply Files/Extensions/LinqExtensions.cs:28]: It is actually used in 5 instances

File: Files/Extensions/EnumerableExtensions.cs [hunk 18]
```diff
@@ -0,0 +1,16 @@
...
+{
+    internal static class EnumerableExtensions
+    {
+        /// <summary>
+        /// Creates <see cref="List{T}"/> and returns <see cref="IEnumerable{T}"/> with provided <paramref name="item"/>
+        /// </summary>
+        /// <typeparam name="T">The item type</typeparam>
+        /// <param name="item">The item</param>
+        /// <returns><see cref="IEnumerable{T}"/> with <paramref name="item"/></returns>
+        internal static IEnumerable<T> CreateEnumerable<T>(this T item) =>
```
[the remark this answers is not in the document]
Comment (d2dyno1) [reply Files/Extensions/EnumerableExtensions.cs:13]: It is actually used in 5 instances

**Comment by d2dyno1** [comment]
:+1:

File: Files/Helpers/Convert/ErrorCodeConverter.cs [hunk 19]
```diff
@@ -0,0 +1,78 @@
...
+                    return ReturnResult.BadArgumentException | ReturnResult.IntegrityCheckFailed;
+
+                case FilesystemErrorCode.ERROR_INPROGRESS:
+                    return ReturnResult.InProgress;
+
+                default: return default;
+            }
+        }
+
+        public static FilesystemErrorCode ToFilesystemErrorCode(this ReturnResult status)
```
[the remark this answers is not in the document]
Comment (d2dyno1) [reply Files/Helpers/Convert/ErrorCodeConverter.cs]: Removed

File: Files/Extensions/EnumExtensions.cs [hunk 20]
```diff
@@ -0,0 +1,25 @@
+﻿using System;
+
+namespace Files.Extensions
+{
+    internal static class EnumExtensions
```
[the remark this answers is not in the document]
Comment (d2dyno1) [reply Files/Extensions/EnumExtensions.cs]: Removed

File: Files/Extensions/LinqExtensions.cs (hunk 16 above)
Comment (d2dyno1) [reply Files/Extensions/LinqExtensions.cs]: Removed

File: Files/Filesystem/StorageFileHelpers/FilesystemResult.cs (hunk 17 above)
[the remark this answers is not in the document]
Comment (tsvietOK) [reply Files/Filesystem/StorageFileHelpers/FilesystemResult.cs:23]: image
Are you sure?

[the remark this answers is not in the document]
Comment (d2dyno1) [reply Files/Filesystem/StorageFileHelpers/FilesystemResult.cs:23]: Oh, I thought you meant the `ERROR_INPROGRESS`. The `FilesystemErrorCodeExtensions.HasFlag()` is obsolete, but it wasn't added in this PR

**Comment by tsvietOK** [comment]
@d2dyno1 
Steps:
1. Create bitmap with any name.
2. Press F5.
3. Try to rename this file

For me it shows:
image

**Comment by tsvietOK** [comment]
@d2dyno1 By the way, when renaming item several times and trying doing undo several times, it reverts only last change.

**Comment by tsvietOK** [comment]
@d2dyno1 Steps:
1. Create folder and place file in this folder.
2. Remove this folder to recycle bin.
3. Press Ctrl+Z.
4. It restore only folder without file inside.

**Comment by d2dyno1** [comment]
@tsvietOK should be working now

File: Files/App.xaml.cs [hunk 21]
```diff
@@ -59,6 +68,30 @@ public App()
...
+            if (history != null)
+            {
+                StorageHistory.Remove(history);
+
+                StorageHistoryIndex--;
+                int index = ArrayHelpers.FitBounds(StorageHistoryIndex, StorageHistory.Count);
+                StorageHistoryIndex = index;
+            }
+        }
+
```
Comment (yaichenbaum) [inline Files/App.xaml.cs:94]: Does this need to be put in app.xaml.cs?

**Review by yair100 — DISMISSED** [review]

File: Files/App.xaml.cs (hunk 21 above)
Comment (d2dyno1) [reply Files/App.xaml.cs:94]: Hmmm :thinking:. I had the only idea to put it here. Where would you move it?

Comment (yaichenbaum) [reply Files/App.xaml.cs:94]: Would it go in a View Model or is there a better way to do it to follow the MVVM pattern?

Comment (duke7553) [reply Files/App.xaml.cs:94]: @d2dyno1 Is this code app-wide or just tied to each tab instance? If it's not the latter, you can place all of this inside a new class file. But please stay away from placing stuff like this inside App.cs 🙂

Comment (d2dyno1) [reply Files/App.xaml.cs]: @duke7553 ok, I wrapped the mess into separate class, it's much cleaner now

**Review by yair100 — DISMISSED** [review]

**Review by lukeblevins — APPROVED** [review]

**Comment by d2dyno1** [comment]
I'd like to thank everyone who helped me with this PR. Without you, this PR wouldn't be as it is now!

Thanks,
@gave92 @tsvietOK @yair100 @duke7553
